### PR TITLE
(docs) Adding an example of a closure that contains mutable state in Mutable State & Imperative Tutorial

### LINKED ISSUE #2035: Add example on closure and references
Page: [Mutability and Imperative Control Flow](https://ocaml.org/docs/mutability-imperative-control-flow)
Source: [0it_05_imperative.md](../blob/main/data/tutorials/language/0it_05_imperative.md)

Insert this example (or something similar) in the imperative tutorial.
```ocaml
let counter () =
  let n = ref 0 in
  fun () -> incr n; !n;;
```

This includes showing how it's used, and how it works.

Credit: https://youtu.be/T_w3OlZ5ASc?t=24m25s

### PR DESCRIPTION
**Pull request by The-Amoghavarsha**
Closes #2035

### CODE REVIEW

File: data/tutorials/language/0it_05_imperative.md [hunk 1]
```diff
@@ -526,6 +526,18 @@ The following example uses the `get_char` function we defined earlier (in the se
...
 This `while` loop echoes characters typed on the keyboard. When the ASCII `Escape` character is read, the `Exit` exception is thrown, which terminates the iteration and displays the REPL reply: `- : unit = ()`.
 
+### Closures
+
+A closure is a function bundled together with its surrounding state, including any variables or references from the enclosing scope that it references. This bundled function-state package allows for the creation of self-contained units of behavior that can maintain internal state across multiple invocations. 
+
+```ocaml
+let counter () =
+  let n = ref 0 in
+  fun () -> incr n; !n;;
```
Comment (cuihtlauac) [inline data/tutorials/language/0it_05_imperative.md]:
```suggestion
# let create_counter () =
  let n = ref 0 in
  fun () -> incr n; !n;;
val create_counter : unit -> unit -> int = <fun>

# let c1 = create_counter ();;
val c1 : unit -> int = <fun>

# let c2 = create_counter ();;
val c2 : unit -> int = <fun>

# c1 ();;
- : int = 1

# c1 ();;
- : int = 2

# c2 ();;
- : int = 1

# c1 ();;
- : int = 3
```

File: data/tutorials/language/0it_05_imperative.md [hunk 2]
```diff
@@ -526,6 +526,18 @@ The following example uses the `get_char` function we defined earlier (in the se
 
 This `while` loop echoes characters typed on the keyboard. When the ASCII `Escape` character is read, the `Exit` exception is thrown, which terminates the iteration and displays the REPL reply: `- : unit = ()`.
 
+### Closures
```
Comment (cuihtlauac) [inline data/tutorials/language/0it_05_imperative.md]:
```suggestion
### References Inside Closures
```

File: data/tutorials/language/0it_05_imperative.md [hunk 3]
```diff
@@ -526,6 +526,18 @@ The following example uses the `get_char` function we defined earlier (in the se
...
+
+A closure is a function bundled together with its surrounding state, including any variables or references from the enclosing scope that it references. This bundled function-state package allows for the creation of self-contained units of behavior that can maintain internal state across multiple invocations.
```
Comment (cuihtlauac) [inline data/tutorials/language/0it_05_imperative.md]: Refer to the definition of closures from [Values & Functios](https://ocaml.org/docs/values-and-functions#closures). 

This example is meant to illustrate it is possible to “hide” a reference in a closure returned by a function. Which is a neat way to produce an "effectful function" (maybe we should say somewhere such a thing is not a all function, in the math sense).

**Review by cuihtlauac — CHANGES_REQUESTED** [review]
Thanks @The-Amoghavarsha. We need more details, this example is more profound that it may looks.

**Comment by cuihtlauac** [comment]
Also, checks are failing because the formatting was broken upstream. Please rebase.

**Review by cuihtlauac — APPROVED** [review]
This looks good to me.
@christinerose, is this ok for you too?

File: data/tutorials/language/0it_05_imperative.md [hunk 4]
```diff
@@ -526,6 +526,44 @@ The following example uses the `get_char` function we defined earlier (in the se
...
+
+In the following example, the function `create_counter` returns a closure that hides a mutable reference `n`. This closure captures the environment where `n` is defined and can modify `n` each time it's invoked. The `n` reference is "hidden" within the closure, encapsulating its state.
+
+```ocaml
+# let create_counter () =
+  let n = ref 0 in
+  fun () -> incr n; !n;;
+val create_counter : unit -> unit -> int = <fun>
+```
+First, we define a function named `create_counter` that takes no arguments. Inside `create_counter`, a reference `n` is initialized with the value 0. This reference will hold the state of the counter. Next, we define a closure that takes no arguments (fun () ->). The closure increments the value of `n` (the counter) using `incr n`, then returns the current value of `n` using `!n`.
```
Comment (christinerose) [inline data/tutorials/language/0it_05_imperative.md]:
```suggestion
First, we define a function named `create_counter` that takes no arguments. Inside `create_counter`, a reference `n` is initialised with the value 0. This reference will hold the state of the counter. Next, we define a closure that takes no arguments (fun () ->). The closure increments the value of `n` (the counter) using `incr n`, then returns the current value of `n` using `!n`.
```

**Review by christinerose — APPROVED** [review]
LGTM!!

**Review by sabine — APPROVED** [review]

**Pull request merged**